### ISSUE TITLE
Corrupt bitstream: multi-pass encoding

### ISSUE THREAD

**Issue by mikoim** [xiph/rav1e#2311]
I encoded video with target bitrate mode to compare quality and performance. rav1e seems to output corrupt bitstream in specific parameters.

```bash
for TILE in 1 2 4 8 16 32 64
do
    for SPEED in 0 1 2 3 4 5 6 7 8 9 10
    do
        echo "--speed $SPEED --tiles $TILE"
        rav1e --output /dev/null --bitrate 3000 --first-pass $SPEED-$TILE.stat --speed $SPEED --tiles $TILE input.y4m
        rav1e --output $SPEED-$TILE.ivf --bitrate 3000 --second-pass $SPEED-$TILE.stat --speed $SPEED --tiles $TILE input.y4m
    done
done
```

## Playable?
| tiles/speed 	|      0     	|      1     	|      2     	|      3     	|      4     	|      5     	|      6     	|      7     	|      8     	|      9     	| 10 	|
|:-----------:	|:----------:	|:----------:	|:----------:	|:----------:	|:----------:	|:----------:	|:----------:	|:----------:	|:----------:	|:----------:	|:--:	|
|      1      	|      ✅     	|      ✅     	|      ✅     	|      ✅     	|      ✅     	|      ✅     	|      ✅     	|      ✅     	|      ✅     	|      ✅     	|  ✅ 	|
|      2      	|      ❌     	|      ❌     	|      ❌     	|      ❌     	|      ❌     	|      ❌     	| not tested 	| not tested 	| not tested 	| not tested 	|  ✅ 	|
|      4      	|      ✅     	|      ✅     	|      ✅     	|      ✅     	|      ✅     	|      ✅     	|      ✅     	|      ✅     	|      ✅     	|      ✅     	|  ✅ 	|
|      8      	|      ✅     	|      ✅     	|      ✅     	|      ✅     	|      ✅     	|      ✅     	|      ✅     	|      ✅     	|      ✅     	|      ✅     	|  ✅ 	|
|      16     	|      ✅     	|      ✅     	|      ✅     	|      ✅     	|      ✅     	|      ✅     	|      ✅     	|      ✅     	|      ✅     	|      ✅     	|  ✅ 	|
|      32     	|      ✅     	|      ✅     	|      ✅     	|      ✅     	|      ✅     	|      ✅     	|      ✅     	|      ✅     	|      ✅     	|      ✅     	|  ✅ 	|
|      64     	| not tested 	| not tested 	| not tested 	| not tested 	| not tested 	| not tested 	|      ❌     	|      ❌     	|      ❌     	|      ❌     	|  ✅ 	|

## Environment
* Intel(R) Core(TM) i9-7980XE CPU @ 2.60GHz
* aom 9c38f3b0f18a0d64cbfbf4d28c283c9a44c69667
* rav1e c6bf0bf76d98dd1bfd943a0723e33bbc40d303a9
    * rustc 1.43.1 (8d69840ab 2020-05-04)
    * ```RUSTFLAGS="-C target-cpu=native" cargo build --release```

```bash
$ mediainfo input.y4m
General
Complete name                            : input.y4m
Format                                   : YUV4MPEG2
File size                                : 451 MiB
Duration                                 : 7 s 958 ms
Overall bit rate                         : 475 Mb/s

Video
Format                                   : YUV
Duration                                 : 7 s 958 ms
Bit rate                                 : 475 Mb/s
Width                                    : 1 718 pixels
Height                                   : 720 pixels
Display aspect ratio                     : 2.40:1
Frame rate                               : 24.000 FPS
Color space                              : YUV
Chroma subsampling                       : 4:2:2
Scan type                                : Progressive
Compression mode                         : Lossless
Bits/(Pixel*Frame)                       : 16.000
Stream size                              : 451 MiB (100%)
```

## Artifacts
* input.y4m: https://gofile.io/d/FqxkFN (cropped from [Cosmos Laundromat: First Cycle](https://cloud.blender.org/p/cosmos-laundromat/))
* corrupt bitstreams: https://gofile.io/d/cmczR5

## Logs
```bash
$ aomdec --verbose -o /dev/null 0-2.ivf
AOMedia Project AV1 Decoder 1.0.0-errata1-avif-893-g9c38f3b0f
Warning: Failed to decode frame 1: Corrupt frame detected
Warning: Additional information: Corrupted segment_ids
```

```bash
$ aomdec --verbose -o /dev/null 1-2.ivf
AOMedia Project AV1 Decoder 1.0.0-errata1-avif-893-g9c38f3b0f
Warning: Failed to decode frame 1: Corrupt frame detected
Warning: Additional information: Block size 8x16 invalid with this subsampling mode
```

```bash
$ aomdec --verbose -o /dev/null 2-2.ivf
AOMedia Project AV1 Decoder 1.0.0-errata1-avif-893-g9c38f3b0f
Warning: Failed to decode frame 1: Corrupt frame detected
Warning: Additional information: Block size 8x32 invalid with this subsampling mode
```

```bash
$ aomdec --verbose -o /dev/null 3-2.ivf
AOMedia Project AV1 Decoder 1.0.0-errata1-avif-893-g9c38f3b0f
Warning: Failed to decode frame 1: Corrupt frame detected
Warning: Additional information: Block size 8x16 invalid with this subsampling mode
```

```bash
$ aomdec --verbose -o /dev/null 4-2.ivf
AOMedia Project AV1 Decoder 1.0.0-errata1-avif-893-g9c38f3b0f
Warning: Failed to decode frame 1: Corrupt frame detected
Warning: Additional information: Block size 16x32 invalid with this subsampling mode
```

```bash
$ aomdec --verbose -o /dev/null 5-2.ivf
AOMedia Project AV1 Decoder 1.0.0-errata1-avif-893-g9c38f3b0f
Warning: Failed to decode frame 1: Corrupt frame detected
Warning: Additional information: Block size 4x16 invalid with this subsampling mode
```

```bash
$ aomdec --verbose -o /dev/null 6-64.ivf
AOMedia Project AV1 Decoder 1.0.0-errata1-avif-893-g9c38f3b0f
Warning: Failed to decode frame 1: Corrupt frame detected
Warning: Additional information: Block size 16x32 invalid with this subsampling mode
```

```bash
$ aomdec --verbose -o /dev/null 7-64.ivf
AOMedia Project AV1 Decoder 1.0.0-errata1-avif-893-g9c38f3b0f
Warning: Failed to decode frame 1: Corrupt frame detected
Warning: Additional information: Block size 16x32 invalid with this subsampling mode
```

```bash
$ aomdec --verbose -o /dev/null 8-64.ivf
AOMedia Project AV1 Decoder 1.0.0-errata1-avif-893-g9c38f3b0f
Warning: Failed to decode frame 1: Corrupt frame detected
Warning: Additional information: Block size 16x32 invalid with this subsampling mode
```

```bash
$ aomdec --verbose -o /dev/null 9-64.ivf
AOMedia Project AV1 Decoder 1.0.0-errata1-avif-893-g9c38f3b0f
Warning: Failed to decode frame 1: Corrupt frame detected
Warning: Additional information: Block size 32x64 invalid with this subsampling mode
```

**Comment by barrbrain** [comment]
Thank you for this detailed report. Which revision of rav1e was used in these tests?

**Comment by mikoim** [comment]
rav1e c6bf0bf76d98dd1bfd943a0723e33bbc40d303a9 is used in these tests.

**Comment by barrbrain** [comment]
I reproduced the 2-tile failures in a test case and added an assertion for the block size error condition. The assertion didn't fail although the decode did, so I hypothesise that the preceding symbols are where the desync occurs. To test this, I tried disabling loop restoration filters (LRF) and the test case then passed. The cause is likely in the writing of LRF coefficients. Given that there is no desync for most resolutions/tiles/subsamplings, it might be that in this case the wrong number of symbols are written.

**Comment by barrbrain** [comment]
I can reproduce in single-pass with quantizers >= 161.

**Comment by mikoim** [comment]
* --speed 9 --tiles 64 --quantizer 176
```
aomdec -o /dev/null s9-t64-q176.ivf
Warning: Failed to decode frame 1: Corrupt frame detected
Warning: Additional information: Failed to decode tile data
```
* --speed 9 --tiles 64 --quantizer 192
```
aomdec -o /dev/null s9-t64-q192.ivf
Warning: Failed to decode frame 1: Corrupt frame detected
Warning: Additional information: Block size 16x32 invalid with this subsampling mode
```
* --speed 9 --tiles 64 --quantizer 208
```
aomdec -o /dev/null s9-t64-q208.ivf
Warning: Failed to decode frame 1: Corrupt frame detected
Warning: Additional information: Block size 16x32 invalid with this subsampling mode
```
* --speed 9 --tiles 64 --quantizer 224
```
aomdec -o /dev/null s9-t64-q224.ivf
Warning: Failed to decode frame 1: Corrupt frame detected
Warning: Additional information: Block size 16x32 invalid with this subsampling mode
```
* --speed 9 --tiles 64 --quantizer 240
```
aomdec -o /dev/null s9-t64-q240.ivf
Warning: Failed to decode frame 1: Corrupt frame detected
Warning: Additional information: Block size 32x64 invalid with this subsampling mode
```

rav1e 02018bfa44edb2ce0b782278c869d6d8011ca8e6
corrupt.tar: https://gofile.io/d/EGd91F

**Comment by xiphmont** [comment]
I think I know what this is; I missed a necessary tile/LRU alignment restriction case for 4:2:2 when fixing #2212.

**Comment by xiphmont** [comment]
Proposed fix in PR #2371, please test.  Our CI coverage is very light on 4:2:2 content right now.

**Comment by mikoim** [comment]
@xiphmont It seems to work well. Thank you!

* sample
s0-t2.ivf, s9-t64.ivf
https://gofile.io/d/txFWIV
